src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
-				
-					assert(iBinding.memory->isCurrentlyMapped());
-				}
+					MDI_SIZE_TYPE offset = INVALID_ADDRESS,
+					multiAllocationSize = {};
+				} requestState;
 
-				auto* vertex_ptr = static_cast<ImDrawVert*>(vBinding.memory->getMappedPointer());
-				auto* index_ptr = static_cast<ImDrawIdx*>(iBinding.memory->getMappedPointer());
+				const auto start = std::chrono::steady_clock::now();
 
-				for (int n = 0; n < drawData->CmdListsCount; n++)
+				//! we must upload entire MDI buffer data to our streaming buffer, but we cannot guarantee allocation can be done in single request
+				for (MDI_SIZE_TYPE uploadedSize = 0ull; uploadedSize < mdiParams.totalByteSizeRequest;)
 				{
-					const ImDrawList* cmd = drawData->CmdLists[n];
-					::memcpy(vertex_ptr, cmd->VtxBuffer.Data, cmd->VtxBuffer.Size * sizeof(ImDrawVert));
-					::memcpy(index_ptr, cmd->IdxBuffer.Data, cmd->IdxBuffer.Size * sizeof(ImDrawIdx));
-					vertex_ptr += cmd->VtxBuffer.Size;
-					index_ptr += cmd->IdxBuffer.Size;
-				}
+					auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start);
+
+					if (elapsed.count() >= 1u)
+					{
+						streamingBuffer->cull_frees();
+						return false;
+					}
+
+					requestState.offset = INVALID_ADDRESS;
+					requestState.multiAllocationSize = streamingBuffer->max_size(); // request available block memory size which we can try to allocate
+
+					static constexpr auto STREAMING_ALLOCATION_COUNT = 1u;
+					const size_t unallocatedSize = m_mdi.buffer->multi_allocate(std::chrono::steady_clock::now() + std::chrono::microseconds(100u), STREAMING_ALLOCATION_COUNT, &requestState.offset, &requestState.multiAllocationSize, &MDI_MAX_ALIGNMENT); //! (*) note we request single tight chunk of memory with max alignment instead of MDI::E_BUFFER_CONTENT separate chunks 
 
-				vBinding.memory->unmap();
-				iBinding.memory->unmap();
+					if (requestState.offset == INVALID_ADDRESS)
+						continue; // failed? lets try again, TODO: should I here have my "blockMemoryFactor =* 0.5" and apply to requestState.multiAllocationSize?
+					else
+					{
+						static constexpr auto ALIGN_OFFSET_NEEDED = 0u;
+						MDI::SUBALLOCATOR_TRAITS_T::allocator_type fillSubAllocator(mdiData, requestState.offset, ALIGN_OFFSET_NEEDED, MDI_MAX_ALIGNMENT, requestState.multiAllocationSize); //! (*) we create linear suballocator to fill the chunk memory (some of at least) with MDI::E_BUFFER_CONTENT data 
+
+						std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> offsets;
+						std::fill(offsets.data(), offsets.data() + MDI_COMPONENT_COUNT, INVALID_ADDRESS);
+						MDI::SUBALLOCATOR_TRAITS_T::multi_alloc_addr(fillSubAllocator, offsets.size(), offsets.data(), mdiParams.bytesToFill.data(), MDI_ALIGNMENTS.data()); //! (*) we suballocate memory regions from the allocated chunk with required alignment per MDI::E_BUFFER_CONTENT block
```

Comment:
you're trying to allocate the whole thing from the linear allocator every time, even is `max_size` is smaller than `totalByteSizeRequest`

you need a different allocation scheme, go through the `drawData->CmdLists` (starting from last command you didn't upload) and do `multi_alloc_addr` call per command list

Reply:
back up an iterator to `drawData->cmdLists` outside the MDI streaming alloc loop, then instead of starting iteration at `n=0` start at `itCmdLists`